### DISCUSSION TITLE
Handling Rate-Limited External API Calls with Concurrent Users in Serverpod

### DISCUSSION THREAD

**Discussion by bteodoru8** [serverpod/serverpod#2821] [Q&A]
I'm trying to make external API calls from a Serverpod-based server. For example, consider a scenario where the server fetches data from an external API (such as the Reddit API), serializes it into a custom model, and returns it to the frontend.

In my project_flutter, I might call an endpoint like client.api.getProfile(token), which then triggers a Dio-based HTTP client on the server to make a GET request to the external API, process the response, and return the data for use in the frontend.

However, my challenge arises when multiple users (e.g., 5 users) interact with this feature at the same time. If the external API enforces a rate limit (for example, 2 requests per second), how can I implement a solution in Serverpod to manage and limit these requests effectively? I’m essentially looking for a way to queue or throttle these requests at the server level, ensuring that the rate limit imposed by the external API is respected across all users.

**Comment by hampuslavin** [comment]
Hi @bteodoru8!
I see what you mean. I don't have a definitive answer but I think you're on the right track 👇 

> I’m essentially looking for a way to queue or throttle these requests at the server level, ensuring that the rate limit imposed by the external API is respected across all users.

To achieve this for a simpler case, you could try adding a [custom interceptor](https://pub-web.flutter-io.cn/packages/dio#interceptors) to the `Dio` client that can check and keep track of the the current request count in `onRequest` (or just wrap the http client in your own rate limiter client). If the max request ceiling is hit you can then delay it perhaps 100 ms and check again.

This would not work/scale if you have more than one Serverpod instance or if your service gets a lot of traffic, but might work depending on the use case!

For more advanced use cases you could use Redis, which as you probably know is integrated into Serverpod by default. Here is a good introduction from the Redis team on this topic: https://redis.io/glossary/rate-limiting/

Hope you can find a way that solves your problem!
